### ISSUE TITLE
wrong dependencies for version 12.1.0 and 12.0.0?

### ISSUE THREAD

**Issue by sheds4tw** [voxpupuli/puppet-r10k#635]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 8.2.1
- Ruby:  ruby 2.5.9p229 (2021-04-05 revision 67939) [x86_64-linux]
- Distribution: RHEL 8
- Module version: 12.1.0 and 12.0.0

## How to reproduce (e.g Puppet code you use)

use versions 12.1.0 and 12.0.0 with puppetlabs-stb pre version 9

## What are you seeing

Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Unknown function: 'stdlib::to_yaml'. (file: /etc/puppetlabs/code/environments/production/modules/r10k/manifests/config.pp, line: 63, column: 16) on node puppet-v3.core.umbc.edu

Looks like you need version 9+ of puppetlabs-stdlib for anything after version 11.0.1 of puppet-r10k

## What behaviour did you expect instead

I expect the readme to indicate you need stdlib 9+ for versions 12.1.0 and 12.0.0 under dependencies

## Any additional information you'd like to impart

Thanks for your time

**Comment by smortex** [comment]
You are right, the minimum version bump for stdlib was forgotten in metadata.json in #613 (breaking change released in 12.0.0) and added in #620 (released in 12.1.1).

Versions 12.0.0 and 12.1.0 are therefore broken for users of an old version of stdlib.  Following semver, when we saw the error:
1. the broken release where not re-rolled (per https://semver.org/#spec-item-3);
2. the fix has been shipped in a new patch release (per https://semver.org/#spec-item-6).

So version 12.1.1 was the first 12.x release with correct minimum version for stdlib in the metadata file.  Users of a supported version had no trouble and experienced no change. Users of an unsupported version had run-time failures and are now notified earlier of the dependency error.

As we only "support" the latest version of the module, we do not have "known issues" / "compatibility matrix" in place in the README for old versions.  Maybe we should yank these known bad releases?

**Comment by sheds4tw** [comment]
This would only be a change to the metadata so maybe re-rolling the old versions is actually reasonable? Probably a spirit of semver vs. truly following semver thing. Philosophical discussion aside, leaving everything as is seems fine to me. This probably only hits someone in the very rare case I found myself in.

I am in the early stages of creating a brand new control repo on a new Puppet server. I pulled the latest version of this module, for compatibility with other modules I found I had to downgrade stdlib a bit. I bumped the version of this module down to make it compatible with stdlib and found a few old versions that did not work, but purported to. Should be pretty rare and not a huge deal anyway.

Thank you for the detailed explanation.
